### PR TITLE
feat: gmp example contract

### PR DESCRIPTION
**Pull request by ahramy**
[AXE-6086](https://axelarnetwork.atlassian.net/browse/AXE-6086)

[AXE-6086]: https://axelarnetwork.atlassian.net/browse/AXE-6086?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hydrobeam — APPROVED** [review]

File: contracts/gmp-example/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[dependencies]
+axelar-gas-service = { workspace = true }
+axelar-gateway = { workspace = true }
+axelar-soroban-std = { workspace = true }
+soroban-sdk = { workspace = true }
+
+[dev_dependencies]
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/Cargo.toml:15]: updated

File: contracts/gmp-example/tests/test.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    let contract_id = env.register_contract(None, GmpExample);
+    let client = GmpExampleClient::new(env, &contract_id);
+
+    client.initialize_gmp_example(gateway, gas_service);
+
+    (client, contract_id)
+}
+
+#[test]
+fn test_gmp_example() {
```
Comment (milapsheth) [inline contracts/gmp-example/tests/test.rs:53]: remove `test` prefix

File: contracts/gmp-example/tests/test.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        setup_app(&env, &destination_gateway_id, &destination_gas_service_id);
+
+    // Set cross-chain message params
+    let source_address = source_app.address.to_string();
+    let destination_address = destination_app_id.to_string();
+    let payload: Bytes = BytesN::<20>::random(&env).into();
+    let payload_hash: BytesN<32> = env.crypto().keccak256(&payload).into();
+
+    // Initiate cross-chain contract call
+    log!(env, "Sending message from source to destination");
```
Comment (milapsheth) [inline contracts/gmp-example/tests/test.rs:80]: do we need these logs vs comments?

File: Cargo.toml [hunk 4]
```diff
@@ -18,6 +18,7 @@ axelar-gateway = { version = "^0.1.0", path = "contracts/axelar-gateway" }
 axelar-operators = { version = "^0.1.0", path = "contracts/axelar-operators" }
 interchain-token-service = { version = "^0.1.0", path = "contracts/interchain-token-service" }
 interchain-token = { version = "^0.1.0", path = "contracts/interchain-token" }
+gmp-example = { version = "^0.1.0", path = "contracts/gmp-example" }
```
[the remark this answers is not in the document]
Comment (ahramy) [reply Cargo.toml:21]: updated.

File: contracts/gmp-example/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+axelar-soroban-std = { workspace = true }
+soroban-sdk = { workspace = true }
+
+[dev-dependencies]
+axelar-gateway = { workspace = true, features = ["testutils"] }
+axelar-soroban-std = { workspace = true, features = ["testutils"] }
+soroban-sdk = { workspace = true, features = ["testutils"] }
+
+[features]
+testutils = ["soroban-sdk/testutils", "axelar-soroban-std/testutils"]
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/Cargo.toml:21]: updated.

File: contracts/gmp-example/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "gmp-example"
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/Cargo.toml:2]: make sense, updated

File: contracts/gmp-example/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        env.storage().instance().set(&DataKey::Initialized, &true);
+        env.storage().instance().set(&DataKey::Gateway, &gateway);
+        env.storage()
+            .instance()
+            .set(&DataKey::GasService, &gas_service);
+
+        Ok(())
+    }
+
+    fn gas_service(env: &Env) -> Address {
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/src/contract.rs:63]: make sense, updated.

File: contracts/gmp-example/src/contract.rs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    fn execute(
+        env: Env,
+        source_chain: String,
+        message_id: String,
+        source_address: String,
+        payload: Bytes,
+    ) {
+        Self::validate(
+            env.clone(),
+            source_chain.clone(),
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/src/contract.rs:29]: updated.

File: contracts/gmp-example/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    fn gas_service(env: &Env) -> Address {
+        env.storage().instance().get(&DataKey::GasService).unwrap()
+    }
+
+    pub fn send(
+        env: Env,
+        destination_chain: String,
+        destination_address: String,
+        message: Bytes,
+        token: Token,
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/src/contract.rs:72]: updated.

File: contracts/gmp-example/tests/test.rs (hunk 2 above)
Comment (ahramy) [reply contracts/gmp-example/tests/test.rs:53]: removing test gives an error: `function `test_gmp_example` is never usedrustc`, so can't be moreved

File: contracts/gmp-example/tests/test.rs (hunk 3 above)
Comment (ahramy) [reply contracts/gmp-example/tests/test.rs:80]: not necessary, removed logs

File: contracts/gmp-example/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        let gateway = AxelarGatewayClient::new(&env, &Self::gateway(&env));
+        let gas_service = AxelarGasServiceClient::new(&env, &Self::gas_service(&env));
+
+        gas_service.pay_gas_for_contract_call(
+            &env.current_contract_address(),
+            &destination_chain,
+            &destination_address,
+            &message,
+            &env.current_contract_address(),
+            &token,
```
[the remark this answers is not in the document]
Comment (ahramy) [reply contracts/gmp-example/src/contract.rs:83]: updated

**Pull request merged**